core/preconditioner/schwarz.cpp

```diff
@@ -0,0 +1,208 @@
+                size_type l_num_rows = parameters_.subdomain_sizes[i];
+                auto rspan = gko::span(offset, offset + l_num_rows);
+                const auto b_view = dense_b->create_submatrix(
+                    rspan, gko::span(0, dense_b->get_size()[1]));
+                auto x_view = dense_x->create_submatrix(
+                    rspan, gko::span(0, dense_x->get_size()[1]));
+                this->subdomain_solvers_[i]->apply(dense_alpha, b_view.get(),
+                                                   dense_beta, x_view.get());
+                offset += l_num_rows;
+            }
+        },
+        alpha, b, beta, x);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Schwarz<ValueType, IndexType>::write(mat_data& data) const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Schwarz<ValueType, IndexType>::transpose() const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Schwarz<ValueType, IndexType>::conj_transpose() const
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void Schwarz<ValueType, IndexType>::generate(const LinOp* system_matrix,
+                                             bool skip_sorting)
+{
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+    using csr_type = matrix::Csr<ValueType, IndexType>;
+    const auto exec = this->get_executor();
+    auto csr_mtx =
+        convert_to_with_sorting<csr_type>(exec, system_matrix, skip_sorting);
+    size_type offset = 0;
+    auto mat_size = system_matrix->get_size();
+    std::vector<size_type> subd_sizes;
+    if (parameters_.subdomain_sizes.size() == 0) {
+        subd_sizes = std::vector<size_type>(num_subdomains_,
+                                            mat_size[0] / num_subdomains_);
+        subd_sizes[0] += mat_size[0] - (num_subdomains_ * subd_sizes[1]);
+    } else {
+        subd_sizes = parameters_.subdomain_sizes;
+    }
+    size_type l_num_rows = 0;
+    // TODO Replace with BlockApprox
+    for (size_type i = 0; i < num_subdomains_; ++i) {
+        size_type l_num_rows = subd_sizes[i];
+        auto rspan = gko::span(offset, offset + l_num_rows);
+        auto cspan = gko::span(offset, offset + l_num_rows);
+        subdomain_matrices_.emplace_back(
+            gko::share(csr_mtx->create_submatrix(rspan, cspan)));
+        offset += l_num_rows;
+    }
+    if (parameters_.generated_inner_solvers.size() > 0) {
+        GKO_ASSERT(parameters_.generated_inner_solvers.size() ==
+                   num_subdomains_);
+    } else if (parameters_.inner_solver) {
+        for (size_type i = 0; i < num_subdomains_; ++i) {
+            subdomain_solvers_.emplace_back(
+                parameters_.inner_solver->generate(subdomain_matrices_[i]));
+        }
+    } else {
+        GKO_NOT_IMPLEMENTED;
+    }
+    if (parameters_.coarse_solver.uses_coarse_solver_) {
```

Comment:
Instead of the `uses_coarse_solver_` boolean, can't you rely on whether `coarse_solver_` is nullptr and remove the need for the bool?